### ISSUE TITLE
V4: Unable to target name for styled component in tests, the name is always "ForwardRef"

### ISSUE THREAD

**Issue by maureenwaggl** [styled-components/styled-components#1985]
## Reproduction

https://codesandbox.io/s/yv7l2p76q9

## Steps to reproduce

Write enzyme test for a styled component to ensure the correct components are rendered.

## Expected Behavior

You should be able to target styled components by their name or explicitly set displayName

## Actual Behavior

All styled components have the name `ForwardRef` instead of the name you give it, making it difficult to differentiate between them

Ideally, it would be nice if the functionality was the same as it was before, where it automatically took the name of whatever you named the styled component as. Most of the tests in our test suite use enzyme in this way and it would cause a considerable refactor to go through and either set an explicit displayName to each instance, or start using specific classes and apply them to all affected components.

Is there a better way of handling this situation?

Thank you!

**Comment by probablyup** [comment]
We actually do forward the displayName: https://github.com/styled-components/styled-components/blob/develop/src/models/StyledComponent.js#L265

Enzyme must be ignoring it or something

**Comment by probablyup** [comment]
You should be able to target the styled-component like this though right?

```js
const Component = styled.div``

mount(<div><Component /></div>).find(Component)
```

**Comment by maureenwaggl** [comment]
Unfortunately not :/ Using `mount` or `shallow`, I can't call the styled component by name. Shallow just shows the `ForwardRef` name, and mount does `BaseStyledComponent` wrapped in the `ForwardRef`.  I updated the repro link to show what I mean between the two

**Comment by probablyup** [comment]
https://github.com/airbnb/enzyme/issues/1810

**Comment by probablyup** [comment]
@maureenwaggl so I think with react 16.5.1 and newest enzyme, things should be targetable again?

**Comment by maureenwaggl** [comment]
@probablyup Thanks for all your help with this and your fast replies and follow ups! I'll keep an eye out for the latest enzyme release and try upgrading when that's available. Thanks again!

**Comment by probablyup** [comment]
FYI I'm adding this to our test-utils: https://github.com/styled-components/styled-components/pull/2049

**Comment by probablyup** [comment]
It's a little tough to test on codesandbox because it uses CRA, but a non-CRA project with the babel plugin should be able to grab the names in the latest enzyme version I think.

**Comment by maureenwaggl** [comment]
Awesome, thank you! I'll try to upgrade soon and hopefully all will work as expected. Thanks again for all your help and followup with this!

**Comment by MadRabbit** [comment]
ohi, i was wondering if there a way to hack around enzyme for nested styled-components?

```js
const Icon = styled.i`...`;
const BlueIcon = styled(Icon)`color: blue`;

const Thing = () => <BlueIcon />;

mount(<Thing />).find(Icon); // -> returns empty
```

the styled component ID hack doesn't seem to work in this case, because classes seem to be reshuffled

**Comment by probablyup** [comment]
Try using the new enzymeFind test util

**Comment by MadRabbit** [comment]
yup, the problem though is that the original `styledComponentId` gets lost in this case. hence i'm wondering if there is a clever way to get around that. here's the full test case:

```js
import styled from 'styled-components';
import { mount } from 'enzyme';
import { enzymeFind } from 'styled-components/test-utils'

describe('bug', () => {
  it('aint no nothing works', () => {
    const Icon = styled.i`color: black`;
    const RedIcon = styled(Icon)`color: red`;
    const BlueIcon = styled(Icon)`color: blue`;

    const TestComponent = () => (
      <div>
        <RedIcon />
        <BlueIcon />
      </div>
    );

    const wrapper = mount(<TestComponent />);

    console.log(wrapper.find(Icon).length);
    console.log(enzymeFind(wrapper, Icon).length);

    console.log(wrapper.find(BlueIcon).length);
  });
});```

it spits out:

```
0
0
1
```

**Comment by MadRabbit** [comment]
hmm, another interesting consequence of this is that `createGlobalStyles` doesn't seem to work with base components either now:

```js
const Sidebar = styled.aside`.....`;
const LeftSidebar = styled(Sidebar)`....`;
const RightSidebar = styled(Sidebar)`...`;

const GlobalStyles = createGlobalStyle`
  ${Sidebar} {
      /* doesn't do anything */
  }

  ${LeftSidebar}, ${RightSidebar} {
     /* works */
  }
`;
```

is that intentional?

**Comment by wxu-aa** [comment]
Any progress on the Styled Component display Name? Or will enzyme fix this issue?
Always show `<StyledComponent>` now. So a lots of Jest tests failed.

Thanks

**Comment by probablyup** [comment]
We’ve done all we can on our side

**Comment by damassi** [comment]
Finding by `displayName` seems to be fixed in the latest enzyme + enzyme-adaptor-react-16.

**Comment by xunuoi** [comment]
@damassi 

Hi, I tried the latest version, but still got something like `<StyledComponent ...` instead of `<MyComponentName ...`?

Could you help explain how you resolve the displayName issue? 

Thanks!!

**Comment by damassi** [comment]
This is separate from shallow mounting, but installing latest allowed me to target by displayName -- meaning, `find('MyComponentName')`. In previous versions + styled-components 4 it would fail.

**Comment by xunuoi** [comment]
Um, seems the issue is still not resolved well, I attached some relevant issue here

https://github.com/airbnb/enzyme/issues/1810 

https://github.com/airbnb/enzyme/issues/1896

https://github.com/airbnb/enzyme/issues/1978
